### PR TITLE
antispoofing project

### PR DESCRIPTION
**Pull request by kprokofi**
Kindly asking for code review
@Ilya-Krylov 
@druzhkov-paul

### CODE REVIEW

File: pytorch_toolkit/light-weight-face-anti-spoofing/.gitignore [hunk 1]
```diff
@@ -0,0 +1,138 @@
+# Byte-compiled / optimized / DLL files
```
Comment (sovrasov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/.gitignore:1]: Please delete this file

File: pytorch_toolkit/light-weight-face-anti-spoofing/README.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+* **external** - parameters for constructing external dataset reader. See Data Preparation section.
+* **img_norm_cfg** - parameters for data normalization
+* **scheduler** - scheduler for dropping learning rate
+* **data.sampler** - if it is true, then will be generated weights for `WeightedRandomSampler` object to uniform distribution of two classes
+* **resize** - resize of the image
+* **checkpoint** - the name of the checkpoint to save and the path to the experiment folder where checkpoint, tensorboard logs and eval metrics will be kept
+* **loss** - there are available two possible losses: `amsoftmax` with `cos`, `arcos`, `cross_enropy` margins and `soft_triple` with different number of inner classes. For more details about this soft triple loss see in [paper](https://arxiv.org/pdf/1909.05235.pdf)
+* **loss.amsoftmax.ratio**a  - there is availability to use different m for different classes. The ratio is the weights on which provided `m` will be divided for a specific class. For example ratio = [1,2] means that m for the first class will equal to m, but for the second will equal to m/2
```
Comment (morkovka1337) [inline pytorch_toolkit/light-weight-face-anti-spoofing/README.md:67]:
```suggestion
* **loss.amsoftmax.ratio**  - there is availability to use different m for different classes. The ratio is the weights on which provided `m` will be divided for a specific class. For example ratio = [1,2] means that m for the first class will equal to m, but for the second will equal to m/2
```
Probably a typo

File: pytorch_toolkit/light-weight-face-anti-spoofing/README.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+* **loss.amsoftmax.gamma** - if this constant differs from 0 then the focal loss will be switched on with the corresponding gamma
+* **For soft triple loss**: `Cn` - number of classes, `K` - number of proxies for each class, `tau` - parameter for regularisation number of proxies
```
Comment (morkovka1337) [inline pytorch_toolkit/light-weight-face-anti-spoofing/README.md:69]:
```suggestion
* **For soft triple loss**: `Cn` - number of classes, `K` - number of proxies for each class, `tau` - parameter for regularization number of proxies
```

File: pytorch_toolkit/light-weight-face-anti-spoofing/README.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+* **model** - there are parameters concern model. `pretrained` means that you want to train with the imagenet weights (you can download weights from [google drive](https://drive.google.com/drive/u/0/folders/1A6wa3AlrdjyNPkXT81knIzXxR7SAYm1q) and specify the path to it in the `imagenet weights` parameter. **model_type** - type of the model, 'Mobilenet3' and 'Mobilenet2' are available. **size** param means the size of the mobilenetv3, there are 'large' and 'small' options. Note that this will change mobilenev3 only. **embeding_dim** - the size of the embeding (vector of features after average pooling). **width_mult** - the width scaling parameter of the model. Note, that you will need the appropriate imagenet weights if you want to train your model with transfer learning. On google drive weights with 0.75, 1.0 value of this parameter is available
```
Comment (morkovka1337) [inline pytorch_toolkit/light-weight-face-anti-spoofing/README.md:70]:
```suggestion
* **model** - there are parameters concern model. `pretrained` means that you want to train with the imagenet weights (you can download weights from [google drive](https://drive.google.com/drive/u/0/folders/1A6wa3AlrdjyNPkXT81knIzXxR7SAYm1q) and specify the path to it in the `imagenet weights` parameter. **model_type** - type of the model, 'Mobilenet3' and 'Mobilenet2' are available. **size** param means the size of the mobilenetv3, there are 'large' and 'small' options. Note that this will change mobilenev3 only. **embeding_dim** - the size of the embeding (vector of features after average pooling). **width_mult** - the width scaling parameter of the model. Note, that you will need the appropriate imagenet weights if you want to train your model with transfer learning. On google drive weights with 0.75, 1.0 value of this parameter are available
```

File: pytorch_toolkit/light-weight-face-anti-spoofing/README.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+* **aug** - advanced augmentation, appropriate value for type is 'cutmix' or 'mixup. lambda = BetaDistribution(alpha, beta), cutmix_prob - probability of applying cutmix on image.
```
Comment (morkovka1337) [inline pytorch_toolkit/light-weight-face-anti-spoofing/README.md:71]:
```suggestion
* **aug** - advanced augmentation, appropriate value for type is 'cutmix' or 'mixup'. lambda = BetaDistribution(alpha, beta), cutmix_prob - probability of applying cutmix on image.
```
Also, consider paraphrasing this sentence, as it may be hard to understand

**Comment by druzhkov-paul** [comment]
@morkovka1337 please do a review in one run — not separate comments!

File: pytorch_toolkit/light-weight-face-anti-spoofing/datasets/casia_surf.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+class CasiaSurfDataset(Dataset):
+    def __init__(self, protocol: int, dir_: str = 'data/CASIA_SURF', mode: str = 'train', depth=False, ir=False,
+                 transform=None):
+        self.dir = dir_
+        self.mode = mode
+        submode = {'train': 'train', 'dev': 'dev_ref',
+                   'test': 'test_res'}[mode]
+        file_name = '4@{}_{}.txt'.format(protocol, submode)
+        with open(os.path.join(dir_, file_name), 'r') as file:
+            lines = file.readlines()
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/datasets/casia_surf.py:32]: Do not load full file into a buffer - use generator: `for line in file`.

File: pytorch_toolkit/light-weight-face-anti-spoofing/datasets/casia_surf.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import torch
+from torch.utils.data import Dataset
+
+
+class CasiaSurfDataset(Dataset):
+    def __init__(self, protocol: int, dir_: str = 'data/CASIA_SURF', mode: str = 'train', depth=False, ir=False,
+                 transform=None):
+        self.dir = dir_
+        self.mode = mode
+        submode = {'train': 'train', 'dev': 'dev_ref',
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/datasets/casia_surf.py:28]: Move `dict` definition to the class level: `class CasiaSurfDataset(Dataset): submodes = {...} def __init__() ...`

File: pytorch_toolkit/light-weight-face-anti-spoofing/datasets/lcc_fasd.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        elif protocol == 'val':
+            spoof_img, real_img = self.get_val_img(self.root_dir)
+            self.list_img = spoof_img + real_img
+            self.labels = torch.cat((torch.ones(len(spoof_img), dtype=torch.long),
+                                    (torch.zeros(len(real_img), dtype=torch.long))))
+        elif protocol == 'test':
+            spoof_img, real_img = self.get_test_img(self.root_dir)
+            self.list_img = spoof_img + real_img
+            self.labels = torch.cat((torch.ones(len(spoof_img), dtype=torch.long),
+                                        (torch.zeros(len(real_img), dtype=torch.long))))
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/datasets/lcc_fasd.py:42]: indent

File: pytorch_toolkit/light-weight-face-anti-spoofing/datasets/lcc_fasd.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                                    + len(spoof_img_val), dtype=torch.long),
+                                    (torch.zeros(len(real_img_train)
+                                    + len(real_img_val), dtype=torch.long))))
+        elif protocol == 'val_test':
+            spoof_img_test, real_img_test = self.get_test_img(self.root_dir)
+            spoof_img_val, real_img_val = self.get_val_img(self.root_dir)
+            self.list_img = spoof_img_test + spoof_img_val + real_img_test + real_img_val
+            self.labels = torch.cat((torch.ones(len(spoof_img_test)
+                                                + len(spoof_img_val), dtype=torch.long),
+                                        (torch.zeros(len(real_img_test)
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/datasets/lcc_fasd.py:57]: indent

File: pytorch_toolkit/light-weight-face-anti-spoofing/datasets/lcc_fasd.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                                        + len(real_img_val), dtype=torch.long))))
+        else:
+            spoof_img_train, real_img_train = self.get_train_img(self.root_dir)
+            spoof_img_val, real_img_val =self. get_val_img(self.root_dir)
+            spoof_img_test, real_img_test = self.get_test_img(self.root_dir)
+            self.list_img = (spoof_img_train + spoof_img_val + spoof_img_test
+                                + real_img_train + real_img_val + real_img_test)
+            self.labels = torch.cat((torch.ones(len(spoof_img_train) + len(spoof_img_val)
+                                                + len(spoof_img_test), dtype=torch.long),
+                                        (torch.zeros(len(real_img_train) + len(real_img_val)
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/datasets/lcc_fasd.py:67]: indent

File: pytorch_toolkit/light-weight-face-anti-spoofing/datasets/lcc_fasd.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        image = cv.imread(img_path, flags=1)
+        if self.get_img_path:
+            return image, img_path
+        image = cv.cvtColor(image, cv.COLOR_BGR2RGB)
+        y_label = self.labels[index]
+        if self.transform:
+            image = self.transform(label=y_label, img=image)['image']
+        # [batch, channels, height, width]
+        image = np.transpose(image, (2, 0, 1)).astype(np.float32)
+        return (torch.tensor(image), y_label)
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/datasets/lcc_fasd.py:83]: `return torch.tensor(image), y_label`

File: pytorch_toolkit/light-weight-face-anti-spoofing/demo_tools/wrapers.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+    def forward(self, batch):
+        batch = self.preprocessing(batch)
+        self.model.eval()
+        model1 = (self.model.module
+                  if self.config.data_parallel.use_parallel
+                  else self.model)
+        with torch.no_grad():
+            output = model1.forward_to_onnx(batch)
+            return output.detach().numpy()
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/demo_tools/wrapers.py:113]: no new line

File: pytorch_toolkit/light-weight-face-anti-spoofing/eval_protocol.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    # load snapshot
+    path_to_experiment = os.path.join(config.checkpoint.experiment_path, config.checkpoint.snapshot_name)
+    epoch_of_checkpoint = load_checkpoint(path_to_experiment, model, map_location=device, optimizer=None)
+
+    # preprocessing, making dataset and loader
+    normalize = A.Normalize(**config.img_norm_cfg)
+    test_transform = A.Compose([
+                A.Resize(**config.resize, interpolation=cv.INTER_CUBIC),
+                normalize
+                ])
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/eval_protocol.py:67]: ugly indent

File: pytorch_toolkit/light-weight-face-anti-spoofing/models/conv2d_cd.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        self.i = 0
+
+    def forward(self, x):
+        out_normal = F.conv2d(input=x, weight=self.weight, bias=self.bias, dilation=self.dilation,
+                              stride=self.stride, padding=self.padding, groups=self.groups)
+        if math.fabs(self.theta - 0.0) < 1e-8:
+            return out_normal
+        else:
+            kernel_diff = self.weight.sum(2).sum(2)
+            kernel_diff = kernel_diff[:, :, None, None]
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/models/conv2d_cd.py:47]: `kernel_diff = self.weight.sum(dim=(2,3), keepdim=True)`?

File: pytorch_toolkit/light-weight-face-anti-spoofing/models/mobilenetv3.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from .conv2d_cd import Conv2d_cd
+from .dropout import Dropout
+
+
+def _make_divisible(v, divisor, min_value=None):
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/models/mobilenetv3.py:24]: Move it some shared file to exclude the duplication between MN-V2 and MN-V3.

File: pytorch_toolkit/light-weight-face-anti-spoofing/models/mobilenetv3.py [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+        output = output.view(output.size(0), -1)
+        spoof_out = self.spoofer(output)
+        if self.multi_heads:
+            type_spoof = self.spoof_type(output)
+            lightning_type = self.lightning(output)
+            real_atr = torch.sigmoid(self.real_atr(output))
+            return spoof_out, type_spoof, lightning_type, real_atr
+        return spoof_out
+
+    def spoof_task(self, features):
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/models/mobilenetv3.py:261]: Methods `make_logits`, `spoof_task`, etc looks similar... Is it better to move them into some parent class?

File: pytorch_toolkit/light-weight-face-anti-spoofing/utils.py [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+    print("\n==> Loading checkpoint")
+    checkpoint = torch.load(checkpoint_path, map_location=map_location)
+    if 'state_dict' in checkpoint:
+        unloaded = net.load_state_dict(checkpoint['state_dict'], strict=strict)
+        missing_keys, unexpected_keys = (', '.join(i) for i in unloaded)
+    else:
+        unloaded = net.load_state_dict(checkpoint, strict=strict)
+        missing_keys, unexpected_keys = (', '.join(i) for i in unloaded)
+    if missing_keys or unexpected_keys:
+        logging.warning(f'NEXT KEYS HAVE NOT BEEN LOADED:\n\nmissing keys: {missing_keys}\
```
Comment (sovrasov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/utils.py:88]: THE FOLLOWING

**Review by sovrasov — CHANGES_REQUESTED** [review]

**Comment by kirillProkofiev** [comment]
Please, review the changed files.

File: pytorch_toolkit/light-weight-face-anti-spoofing/models/mobilenetv2.py [hunk 18]
```diff
@@ -110,85 +62,46 @@ def forward(self, x):
...
-        if self.multi_heads:
-            self.lightning = nn.Linear(embeding_dim, 5)
-            self.spoof_type = nn.Linear(embeding_dim, 11)
-            self.real_atr = nn.Linear(embeding_dim, 40)
+        self.conv_last = conv_1x1_bn(input_channel, self.embeding_dim)
 
     def forward(self, x):
-        x = self.features(x)
-        x = self.conv_last(x)
+        x = super().make_features(x)
```
Comment (evgeny-izutov) [inline pytorch_toolkit/light-weight-face-anti-spoofing/models/mobilenetv2.py:89]: `forward()` methods are identical between MN-V2,-V3. So move it to the parent class too.

**Review by sovrasov — APPROVED** [review]

**Review by evgeny-izutov — APPROVED** [review]
